### PR TITLE
R3SOL-372 Extract the Backing Store and Uniqueness Checker to a library

### PR DESCRIPTION
**Pull request by nkovacsx**
Main changes in this PR:
- `BackingStore` and `UniquenessChecker` interfaces are now part of the `ledger-lib-uniqueness` library
- The `BackingStore` logic is basically common for C5 and DLT except for how we establish the DB connection.
   That's the reason we have a base abstract class now called `JPABackingStoreBase`. 
   Both C5 and DLT will inherit from this. In C5 it's called `JPABackingStoreOsgiImpl` and the entity manager 
   is acquired through `DbConnectionManager`.
- Thankfully the core uniqueness checker logic is the same for C5/DLT so only C5 will extend 
  `BatchedUniquenessCheckerImpl` with the necessary OSGi things.
- `BatchedUniquenessCheckerImplTest` tests the main UniquenessChecker logic
- `JPABackingStoreOsgiImplIntegrationTests`, `JPABackingStoreEntitiesIntegrationTest` tests BackingStore in C5. 
    We will have testcontainers integration tests in DLT.
- `JPABackingStoreOsgiImplBenchmark` still exists and still works.
- `UniquenessCheckerImplDBIntegrationTests` is the integration test for C5, will have testcontainers version in DLT.

### CODE REVIEW

File: components/uniqueness/backing-store-impl/src/main/kotlin/net/corda/uniqueness/backingstore/impl/osgi/JPABackingStoreOsgiImpl.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    private val jpaEntitiesRegistry: JpaEntitiesRegistry,
+    @Reference(service = DbConnectionManager::class)
+    private val dbConnectionManager: DbConnectionManager,
+    @Reference(service = PersistenceExceptionCategorizer::class)
+    private val persistenceExceptionCategorizer: PersistenceExceptionCategorizer,
+    @Reference(service = VirtualNodeInfoReadService::class)
+    private val virtualNodeInfoReadService: VirtualNodeInfoReadService,
+    @Reference(service = BackingStoreMetricsFactory::class)
+    private val backingStoreMetricsFactory: BackingStoreMetricsFactory
+) : JPABackingStoreBase(backingStoreMetricsFactory, persistenceExceptionCategorizer) {
```
Comment (lankydan) [inline components/uniqueness/backing-store-impl/src/main/kotlin/net/corda/uniqueness/backingstore/impl/osgi/JPABackingStoreOsgiImpl.kt]: I'd prefer that we use delegation instead of abstract/open classes.

Something like:

```kotlin
@Component(service = [BackingStore::class])
class JPABackingStoreOsgiImpl(delegate: JPABackingStore) : JPABackingStore by delegate {

    @Activate constructor(
        @Reference(service = JpaEntitiesRegistry::class)
        jpaEntitiesRegistry: JpaEntitiesRegistry,
        @Reference(service = DbConnectionManager::class)
        dbConnectionManager: DbConnectionManager,
        @Reference(service = PersistenceExceptionCategorizer::class)
        persistenceExceptionCategorizer: PersistenceExceptionCategorizer,
        @Reference(service = VirtualNodeInfoReadService::class)
        virtualNodeInfoReadService: VirtualNodeInfoReadService,
        @Reference(service = BackingStoreMetricsFactory::class)
        backingStoreMetricsFactory: BackingStoreMetricsFactory
    ) : this(
        JPABackingStoreImpl(
            getEntityManagerFactory = { getEntityManagerFactory(holdingIdentity)},
            persistenceExceptionCategorizer = persistenceExceptionCategorizer,
            backingStoreMetricsFactory = backingStoreMetricsFactory
        )
    )

    init {
        jpaEntitiesRegistry.register(
            CordaDb.Uniqueness.persistenceUnitName,
            JPABackingStoreEntities.classes
        )
    }

    private fun getEntityManagerFactory(holdingIdentity: HoldingIdentity): EntityManagerFactory {
        val virtualNodeInfo = virtualNodeInfoReadService.getByHoldingIdentityShortHash(holdingIdentity.shortHash)
            ?: throw VirtualNodeNotFoundException("Virtual node ${holdingIdentity.shortHash} not found")
        val uniquenessDmlConnectionId = virtualNodeInfo.uniquenessDmlConnectionId
        requireNotNull(uniquenessDmlConnectionId) { "uniquenessDmlConnectionId is null" }

        return dbConnectionManager.getOrCreateEntityManagerFactory(
            uniquenessDmlConnectionId,
            entitiesSet = jpaEntitiesRegistry.get(CordaDb.Uniqueness.persistenceUnitName)
                ?: throw IllegalStateException(
                    "persistenceUnitName " +
                        "${CordaDb.Uniqueness.persistenceUnitName} is not registered."
                )
        )
    }
}
```

File: components/uniqueness/backing-store/src/main/kotlin/net/corda/uniqueness/backingstore/osgi/BackingStoreLifecycle.kt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package net.corda.uniqueness.backingstore.osgi
+
+import net.corda.ledger.libs.uniqueness.backingstore.BackingStore
+import net.corda.lifecycle.Lifecycle
+
+
+interface BackingStoreLifecycle : BackingStore, Lifecycle
```
Comment (lankydan) [inline components/uniqueness/backing-store/src/main/kotlin/net/corda/uniqueness/backingstore/osgi/BackingStoreLifecycle.kt:7]: Do we need this interface?

File: components/uniqueness/uniqueness-checker-impl/src/integrationTest/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckerImplDBIntegrationTests.kt [hunk 3]
```diff
@@ -318,20 +312,21 @@ class UniquenessCheckerImplDBIntegrationTests {
...
 
     @Nested
     inner class MalformedRequests {
         @Test
+        @Disabled("Setting null is not possible anymore")
         fun `Request is missing time window upper bound`() {
             assertThrows(AvroRuntimeException::class.java, {
                 newRequestBuilder()
-                    .setTimeWindowUpperBound(null)
+                    //.setTimeWindowUpperBound(null)
```
Comment (lankydan) [inline components/uniqueness/uniqueness-checker-impl/src/integrationTest/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckerImplDBIntegrationTests.kt]: Comment.

File: components/uniqueness/uniqueness-checker-impl/src/integrationTest/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckerImplDBIntegrationTests.kt [hunk 4]
```diff
@@ -1929,7 +1924,8 @@ class UniquenessCheckerImplDBIntegrationTests {
 
             // check that the failure gets replayed via the checker
             processRequests(
-                newRequestBuilder(transactionId, UniquenessCheckType.READ)
+                newRequestBuilder(transactionId)
+                    .setCheckType(net.corda.ledger.libs.uniqueness.data.UniquenessCheckType.READ)
```
Comment (lankydan) [inline components/uniqueness/uniqueness-checker-impl/src/integrationTest/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckerImplDBIntegrationTests.kt]: Does this need the fully qualified imports?

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/BatchedUniquenessCheckerOsgiImpl.kt [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import org.osgi.service.component.annotations.Activate
+import org.osgi.service.component.annotations.Component
+import org.osgi.service.component.annotations.Reference
+
+@Component(service = [ UniquenessChecker::class ])
+class BatchedUniquenessCheckerOsgiImpl(
+    backingStore: BackingStore,
+    uniquenessCheckerMetricsFactory: UniquenessCheckerMetricsFactory,
+    clock: Clock
+) : BatchedUniquenessCheckerImpl(backingStore, uniquenessCheckerMetricsFactory, clock) {
```
Comment (lankydan) [inline components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/BatchedUniquenessCheckerOsgiImpl.kt]: Use delegation instead.

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckerAvroUtils.kt [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+import net.corda.data.uniqueness.UniquenessCheckRequestAvro
+import net.corda.data.uniqueness.UniquenessCheckResponseAvro
+import net.corda.data.uniqueness.UniquenessCheckType
+import net.corda.ledger.libs.uniqueness.data.UniquenessCheckRequest
+import net.corda.ledger.libs.uniqueness.data.UniquenessCheckResponse
+import net.corda.uniqueness.datamodel.common.toAvro
+import net.corda.virtualnode.toCorda
+
+object UniquenessCheckerAvroUtils {
```
Comment (lankydan) [inline components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckerAvroUtils.kt:34]: Is there a benefit to having the object to contain the methods?

File: components/uniqueness/uniqueness-checker/src/main/kotlin/net/corda/uniqueness/checker/UniquenessCheckerLifecycle.kt [hunk 7]
```diff
@@ -0,0 +1,6 @@
+package net.corda.uniqueness.checker
+
+import net.corda.ledger.libs.uniqueness.UniquenessChecker
+import net.corda.lifecycle.Lifecycle
+
+interface UniquenessCheckerLifecycle : UniquenessChecker, Lifecycle
```
Comment (lankydan) [inline components/uniqueness/uniqueness-checker/src/main/kotlin/net/corda/uniqueness/checker/UniquenessCheckerLifecycle.kt:6]: Do we need this interface?

File: libs/ledger-lib-uniqueness/build.gradle [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    implementation libs.jackson.datatype.jsr310
+
+    implementation 'net.corda:corda-application'
+    implementation 'net.corda:corda-crypto'
+    implementation "org.hibernate:hibernate-core:$hibernateVersion"
+
+    implementation platform("net.corda:corda-api:$cordaApiVersion")
+
+    // TODO This can be removed once we have our own SecureHash implementation, we only need this for the OpaqueBytes
+    implementation project(":libs:base-internal")
```
Comment (lankydan) [inline libs/ledger-lib-uniqueness/build.gradle]: Is the comment true, since we're in the uniqueness lib, we shouldn't need to manually create secure hashes, if we are then they should be constructed within a lambda or a factory/interface to abstract it out.

File: libs/ledger-lib-uniqueness/src/main/kotlin/net/corda/ledger/libs/uniqueness/backingstore/BackingStore.kt [hunk 9]
```diff
@@ -45,16 +43,16 @@ import net.corda.virtualnode.HoldingIdentity
...
-
 interface BackingStore {
-
     /**
      * Opens a new session with the backing store and runs the supplied block of code in the
      * context of the session. A session is tied to a specific [holdingIdentity], which must be
      * specified when opening the session.
      */
-    fun session(holdingIdentity: HoldingIdentity, block: (Session) -> Unit)
+    fun session(
```
Comment (lankydan) [inline libs/ledger-lib-uniqueness/src/main/kotlin/net/corda/ledger/libs/uniqueness/backingstore/BackingStore.kt]: Does this need reformatting, don't really care though.

File: libs/ledger-lib-uniqueness/src/main/kotlin/net/corda/ledger/libs/uniqueness/backingstore/impl/JPABackingStoreBase.kt [hunk 10]
```diff
@@ -228,21 +185,18 @@ open class JPABackingStoreImpl @Activate constructor(
             existing.forEach { stateEntity ->
                 val consumingTxId =
                     if (stateEntity.consumingTxId != null) {
-                        SecureHashImpl(stateEntity.consumingTxIdAlgo!!, stateEntity.consumingTxId!!)
+                        UniquenessSecureHashImpl(stateEntity.consumingTxIdAlgo!!, stateEntity.consumingTxId!!)
```
Comment (lankydan) [inline libs/ledger-lib-uniqueness/src/main/kotlin/net/corda/ledger/libs/uniqueness/backingstore/impl/JPABackingStoreBase.kt]: Yeah, related to my other comment on creating secure hashes, I think we could have a factory that takes in the inputs to a secure hash and then we can have C5 construct one that creates a `SecureHashImpl` and the DLT can create its own impl.

Although this only matters if we're going to break the dep on `crypto-core` and I'm not yet massively convinced that we do need to break the dependency.

File: libs/ledger-lib-uniqueness/src/main/kotlin/net/corda/ledger/libs/uniqueness/data/UniquenessCheckRequest.kt [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+import net.corda.uniqueness.datamodel.impl.UniquenessCheckStateRefImpl
+import net.corda.uniqueness.datamodel.internal.UniquenessCheckRequestInternal
+import net.corda.v5.application.uniqueness.model.UniquenessCheckStateRef
+import net.corda.v5.base.annotations.CordaSerializable
+import net.corda.v5.base.types.MemberX500Name
+import net.corda.virtualnode.HoldingIdentity
+import java.time.Instant
+
+@CordaSerializable
+enum class UniquenessCheckType {
```
Comment (lankydan) [inline libs/ledger-lib-uniqueness/src/main/kotlin/net/corda/ledger/libs/uniqueness/data/UniquenessCheckRequest.kt]: I'd make this a type inside of `UniquenessCheckRequest` to its `UniquenessCheckRequest.Type`.

File: libs/ledger-lib-uniqueness/src/main/kotlin/net/corda/ledger/libs/uniqueness/data/UniquenessSecureHashImpl.kt [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+import net.corda.v5.base.annotations.CordaSerializable
+import net.corda.v5.base.util.ByteArrays
+import net.corda.v5.crypto.SecureHash
+import net.corda.v5.crypto.SecureHash.DELIMITER
+import java.nio.ByteBuffer
+import java.security.MessageDigest
+
+// TODO This is just a 1-to-1 copy of `SecureHashImpl` and renamed to avoid confusion
+@CordaSerializable
+class UniquenessSecureHashImpl(
```
Comment (lankydan) [inline libs/ledger-lib-uniqueness/src/main/kotlin/net/corda/ledger/libs/uniqueness/data/UniquenessSecureHashImpl.kt]: need to review.

File: libs/ledger-lib-uniqueness/src/main/kotlin/net/corda/ledger/libs/uniqueness/impl/BatchedUniquenessCheckerImpl.kt [hunk 13]
```diff
@@ -120,17 +105,20 @@ class BatchedUniquenessCheckerImpl(
...
+                        Instant.now(),
+                        UniquenessCheckErrorMalformedRequestImpl(e.message ?: "Unknown error.")
+                    )
                 )
                 ++numMalformed
             }
         }
 
-        if ( numMalformed > 0 ) { log.debug { "$numMalformed malformed requests were rejected" } }
+        if ( numMalformed > 0 ) { log.debug ( "$numMalformed malformed requests were rejected" ) }
```
Comment (lankydan) [inline libs/ledger-lib-uniqueness/src/main/kotlin/net/corda/ledger/libs/uniqueness/impl/BatchedUniquenessCheckerImpl.kt]: Why change the log?

File: testing/uniqueness/backing-store-fake/src/main/kotlin/net/corda/uniqueness/backingstore/impl/fake/BackingStoreImplFake.kt [hunk 14]
```diff
@@ -24,32 +15,10 @@ import net.corda.v5.application.uniqueness.model.UniquenessCheckStateDetails
...
-    private companion object {
-        private val log: Logger = LoggerFactory.getLogger(this::class.java.enclosingClass)
-    }
-
-    private val lifecycleCoordinator: LifecycleCoordinator = coordinatorFactory
-        .createCoordinator<BackingStoreLifecycle>(::eventHandler)
 
-    override val isRunning: Boolean
-        get() = lifecycleCoordinator.isRunning
+@Suppress("ForbiddenComment")
```
Comment (lankydan) [inline testing/uniqueness/backing-store-fake/src/main/kotlin/net/corda/uniqueness/backingstore/impl/fake/BackingStoreImplFake.kt:19]: Lol wtf is forbidden comment 🤣

File: testing/uniqueness/backing-store-fake/src/main/kotlin/net/corda/uniqueness/backingstore/impl/fake/BackingStoreImplFake.kt [hunk 15]
```diff
@@ -24,32 +15,10 @@ import net.corda.v5.application.uniqueness.model.UniquenessCheckStateDetails
...
+open class BackingStoreImplFake : BackingStore {
```
Comment (lankydan) [inline testing/uniqueness/backing-store-fake/src/main/kotlin/net/corda/uniqueness/backingstore/impl/fake/BackingStoreImplFake.kt:20]: Why does this class need opening?

File: testing/uniqueness/uniqueness-utilities/build.gradle [hunk 16]
```diff
@@ -22,4 +22,8 @@ dependencies {
     implementation project(":libs:utilities")
     implementation project(":testing:crypto-testkit")
     implementation project(":testing:test-utilities")
+    implementation project(":libs:ledger-lib-uniqueness")
+
+    // TODO Remove this once we work around the holding identities
+    implementation project(":libs:virtual-node:virtual-node-info")
```
Comment (lankydan) [inline testing/uniqueness/uniqueness-utilities/build.gradle:28]: What do you mean here?

**Review by lankydan — COMMENTED** [review]
we should run `./gradlew dlt:dependencies` on the `ledger-lib-persistence` module so we can review the dependencies and transitive dependencies.

File: components/uniqueness/backing-store/src/main/kotlin/net/corda/uniqueness/backingstore/osgi/BackingStoreLifecycle.kt (hunk 2 above)
Comment (nkovacsx) [reply components/uniqueness/backing-store/src/main/kotlin/net/corda/uniqueness/backingstore/osgi/BackingStoreLifecycle.kt:7]: I think OSGi needs it but I'm not sure tbh...

File: testing/uniqueness/backing-store-fake/src/main/kotlin/net/corda/uniqueness/backingstore/impl/fake/BackingStoreImplFake.kt (hunk 14 above)
Comment (nkovacsx) [reply testing/uniqueness/backing-store-fake/src/main/kotlin/net/corda/uniqueness/backingstore/impl/fake/BackingStoreImplFake.kt:19]: I have no idea why this is there lol I can see if I can remove it.

File: testing/uniqueness/uniqueness-utilities/build.gradle (hunk 16 above)
Comment (nkovacsx) [reply testing/uniqueness/uniqueness-utilities/build.gradle:28]: Oops, that was just a mistake.

File: testing/uniqueness/backing-store-fake/src/main/kotlin/net/corda/uniqueness/backingstore/impl/fake/BackingStoreImplFake.kt (hunk 15 above)
Comment (nkovacsx) [reply testing/uniqueness/backing-store-fake/src/main/kotlin/net/corda/uniqueness/backingstore/impl/fake/BackingStoreImplFake.kt:20]: It was open before, I believe because we `spy` it.

File: libs/ledger-lib-uniqueness/src/main/kotlin/net/corda/ledger/libs/uniqueness/impl/BatchedUniquenessCheckerImpl.kt (hunk 13 above)
Comment (nkovacsx) [reply libs/ledger-lib-uniqueness/src/main/kotlin/net/corda/ledger/libs/uniqueness/impl/BatchedUniquenessCheckerImpl.kt]: No idea, reverted it back.

File: components/uniqueness/uniqueness-checker-impl/src/integrationTest/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckerImplDBIntegrationTests.kt (hunk 3 above)
Comment (nkovacsx) [reply components/uniqueness/uniqueness-checker-impl/src/integrationTest/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckerImplDBIntegrationTests.kt]: I had to comment this out because setting `null` on a request is not possible anymore. It's not avro now, that's why this test is `@Disabled("Setting null is not possible anymore")`.

File: components/uniqueness/uniqueness-checker-impl/src/integrationTest/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckerImplDBIntegrationTests.kt (hunk 4 above)
Comment (nkovacsx) [reply components/uniqueness/uniqueness-checker-impl/src/integrationTest/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckerImplDBIntegrationTests.kt]: Good point, removed

File: components/uniqueness/backing-store-impl/src/main/kotlin/net/corda/uniqueness/backingstore/impl/osgi/JPABackingStoreOsgiImpl.kt (hunk 1 above)
Comment (nkovacsx) [reply components/uniqueness/backing-store-impl/src/main/kotlin/net/corda/uniqueness/backingstore/impl/osgi/JPABackingStoreOsgiImpl.kt]: Using delegation now.

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/BatchedUniquenessCheckerOsgiImpl.kt (hunk 5 above)
Comment (nkovacsx) [reply components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/BatchedUniquenessCheckerOsgiImpl.kt]: Done

File: components/uniqueness/uniqueness-checker/src/main/kotlin/net/corda/uniqueness/checker/UniquenessCheckerLifecycle.kt (hunk 7 above)
Comment (nkovacsx) [reply components/uniqueness/uniqueness-checker/src/main/kotlin/net/corda/uniqueness/checker/UniquenessCheckerLifecycle.kt:6]: I think it's needed for OSGi but idk

File: libs/ledger-lib-uniqueness/src/main/kotlin/net/corda/ledger/libs/uniqueness/data/UniquenessCheckRequest.kt (hunk 11 above)
Comment (nkovacsx) [reply libs/ledger-lib-uniqueness/src/main/kotlin/net/corda/ledger/libs/uniqueness/data/UniquenessCheckRequest.kt]: done

**Review by lankydan — DISMISSED** [review]
I'm ok with this.

File: components/uniqueness/backing-store-impl/src/integrationTest/kotlin/net/corda/uniqueness/backingstore/impl/JPABackingStoreOsgiImplIntegrationTests.kt [hunk 17]
```diff
@@ -84,14 +91,16 @@ import kotlin.reflect.full.createInstance
...
+class JPABackingStoreOsgiImplIntegrationTests {
+    private lateinit var backingStoreImpl: JPABackingStoreOsgiImpl
     private lateinit var testClock: AutoTickTestClock
     private val baseTime = Instant.EPOCH
     private val defaultTimeWindowUpperBound = LocalDate.of(2200, 1, 1).atStartOfDay().toInstant(ZoneOffset.UTC)
 
     private val groupId = UUID.randomUUID().toString()
     private val notaryVNodeIdentity = createTestHoldingIdentity("C=GB, L=London, O=NotaryRep1", groupId)
+    private val notaryVNodeUniquenessHoldingIdentity = createTestHoldingIdentity("C=GB, L=London, O=NotaryRep1", groupId)
+        .let { UniquenessHoldingIdentity(it.x500Name, it.groupId, it.shortHash, it.hash) }
```
Comment (blsemo) [inline components/uniqueness/backing-store-impl/src/integrationTest/kotlin/net/corda/uniqueness/backingstore/impl/JPABackingStoreOsgiImplIntegrationTests.kt:103]: Why do we need a `noataryVNodeIdentity` and a `notaryVNodeUniquenessHoldingIdentity` with the same X500 names and groupID? That seems redundant at best, but possibly wrong?

File: components/uniqueness/uniqueness-checker-impl/src/integrationTest/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckerImplDBIntegrationTests.kt (hunk 3 above)
Comment (blsemo) [reply components/uniqueness/uniqueness-checker-impl/src/integrationTest/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckerImplDBIntegrationTests.kt]: If it can't be done, it should be removed - never commit commented out code.

File: libs/ledger-lib-uniqueness/src/main/kotlin/net/corda/ledger/libs/uniqueness/impl/BatchedUniquenessCheckerImpl.kt [hunk 18]
```diff
@@ -388,26 +379,28 @@ class BatchedUniquenessCheckerImpl(
...
-        holdingIdentity: HoldingIdentity,
+        holdingIdentity: UniquenessHoldingIdentity,
         batch: List<UniquenessCheckRequestInternal>
     ): List<Pair<UniquenessCheckRequestInternal, InternalUniquenessCheckResultWithContext>> {
-
         val resultsToRespondWith =
             mutableListOf<Pair<UniquenessCheckRequestInternal, InternalUniquenessCheckResultWithContext>>()
 
-        log.debug { "Processing uniqueness check read batch of ${batch.size} requests for $holdingIdentity" }
+        log.debug("Processing uniqueness check read batch of ${batch.size} requests for $holdingIdentity")
```
Comment (blsemo) [inline libs/ledger-lib-uniqueness/src/main/kotlin/net/corda/ledger/libs/uniqueness/impl/BatchedUniquenessCheckerImpl.kt]: Why the change to the function form from the lambda form here? That has a potential performance impact...

**Review by blsemo — COMMENTED** [review]
I think the overall intent looks good, I've had a couple of minor points.
The main issue is have is the creation of a holding id and secure hash specific to uniqueness -don't we have the exactly same problem in all of the libs?
I don't think we can avoid a refactoring to  have a secure hash impl that works independently of the majority of Corda 5.

**Comment by blsemo** [comment]
So - we could possibly merge this PR with the UniquenessHoldingIdentity and UniquenessSecureHash, and then have follow up tickets to refactor HoldingIdentity and SecureHash to be usable separately, to keep the PRs manageable.

**Review by lankydan — DISMISSED** [review]
remember to retarget feature branch

**Review by lankydan — APPROVED** [review]
